### PR TITLE
feat: docs for data pipelines v2.

### PR DESCRIPTION
**Pull request by ZuhaaJamani**
Thank you for contributing to the documentation.

Do the changes you have made apply to both Current and Previous versions?

Have you checked all new or changed links?

Is there anything else you would like us to know?

For reference: 

  - We have an [internal contributions guide](https://www.notion.so/genesisglobal/Contributing-new-documentation-75953fb245f246ff872789035451a0c4)
  - We have a [style guide](https://www.notion.so/genesisglobal/Documentation-style-guide-5b04ec6fe12f4262b90d192effd8059b) 

**This week's exciting excerpts from the style guide**

- We write in UK English, with Oxford English Dictionary (OED) spellings, grammar and vocabulary.  

Use the present tense wherever possible. Only use another tense where it is strictly necessary.

- Present tense (preferred whenever possible): *The script creates a new folder.*
- Future tense (avoid whenever possible): *The script will create a new folder.*

### CODE REVIEW

File: docs/03_server/09_data-pipelines/02_basics.mdx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+- Batch Pipeline
+- Realtime Pipeline
+
+Each pipeline must have a single source and single sink, in between these stages you are able to have any number of operator steps as long as the types & ordering comply to that of the pipeline you are using.
+See below for more details on what types are available among the 2 pipelines.
+
+### Batch Pipeline
+
+A batch pipeline will only run once and will need to be triggered manually via the use of its execute method inside the Pipeline Manager.
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/02_basics.mdx:58]:
```suggestion
A batch pipeline will only run once and will need to be triggered manually via the use of its execute method. The pipeline can be retrieved from the Pipeline Manager class.
```

File: docs/03_server/09_data-pipelines/03_implementations.mdx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+```
+
+### Database Batch Poll Source (Realtime)
+
+Realtime acknowledgement source that returns a stream of updates via a batch poll on a particular database dependent on an acknowledgement having been received of the previous element. If no data is received, the source will wait for a delay before attempting to poll again. This source implementation uses the PersistenceManager to persist the latest index using the provided lambda - by default the DbPersistenceManager is used to manage the state of this source. See Persistence Manager for more details (link to persistence manager).
+
+The type of object that you would like this source to return must be given as part of the source definition, example shown below.
+
+When configuring the db batch poller to use as part of your pipeline you must provide the following as part of the configuration:
+- source (required): this is what the PersistenceManager will use to find the last persisted record and should be should be unique across all pipelines
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/03_implementations.mdx:36]:
```suggestion
- source (required): this is what the PersistenceManager will use to find the last persisted record and should be unique across all pipelines
```

File: docs/03_server/09_data-pipelines/03_implementations.mdx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+## Operators
+
+### CSV Decoder
+
+Split Operator to decode from an InputStream (from a PipelineFile object) to a CsvRow object which contains:
+
+1. the line number for each CSV row
+2. an object of a user defined type representing the data of each CSV line, by default this is of type `Map<String, String>`
+
+GPAL functions available are `csvRawDecoder()` to deserialise to `Map<String, String>` and `csvDecoder<Trade>()` to deserialise to specified class.
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/03_implementations.mdx:73]:
```suggestion
GPAL functions available are `csvRawDecoder()` to deserialise to `Map<String, String>` and `csvDecoder<T>()` to deserialise to the specified class, represented by T here.
```

File: docs/03_server/09_data-pipelines/03_implementations.mdx [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+GPAL Examples:
+```kotlin
+// Decoding to Map<String, String>
+pipeline("TEST_PIPELINE_RAW_DECODER") {
+    source(camelSource {
+        location = ""
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/03_implementations.mdx:80]:
```suggestion
        location = "file://src/test/resources/camel-source-test?noop=true&idempotent=true"
```

File: docs/03_server/09_data-pipelines/03_implementations.mdx [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                }
+            )
+        }
+        .sink(txDbSink())
+}
+
+// Decoding to another class
+pipeline("TEST_PIPELINE_DECODER") {
+    source(camelSource {
+        location = ""
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/03_implementations.mdx:96]:
```suggestion
        location = "file://src/test/resources/camel-source-test?noop=true&idempotent=true"
```

File: docs/03_server/09_data-pipelines/03_implementations.mdx (hunk 4 above)
Comment (khouari1) [reply docs/03_server/09_data-pipelines/03_implementations.mdx:80]: Just to be clear that an empty string is not a valid example (may not be required to point out but you never know!)

File: docs/03_server/09_data-pipelines/03_implementations.mdx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+// User Defines SimpleTrade Class for type of each CSV Row
+pipeline("TEST_CSV_DECODER") {
+        source(CsvDecoderTestSource(File(getResourcePath("/Trades.csv"))))
+            .split(CsvDecoder(SimpleTrade::class.java))
+            .sink(CsvDecoderTestSink())
+    }
+
+```
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/03_implementations.mdx:111]: This is unnecessary as we have provided an example of using a custom class above

File: docs/03_server/09_data-pipelines/03_implementations.mdx [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    source(dbBulkSubscribe<Trade>())
+        .map {
+            val record = when (it) {
+                Bulk.Prime.Completed -> null
+                is Bulk.Prime.Record -> it.record
+                is Bulk.Update.Delete -> null
+                is Bulk.Update.Insert -> it.record
+                is Bulk.Update.Modify -> it.record
+            }
+            DbOperation.Insert(record!!)
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/03_implementations.mdx:137]:
```suggestion
            when (it) {
                    Bulk.Prime.Completed -> null
                    is Bulk.Prime.Record -> null
                    is Bulk.Update.Delete -> DbOperation.Delete(it.record)
                    is Bulk.Update.Insert -> DbOperation.Insert(it.record)
                    is Bulk.Update.Modify -> DbOperation.Modify(it.record)
                }
```

File: docs/03_server/09_data-pipelines/03_implementations.mdx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  source(dbBulkSubscribe<Trade>())
+      .map {
+          val record = when (it) {
+              Bulk.Prime.Completed -> null
+              is Bulk.Prime.Record -> it.record
+              is Bulk.Update.Delete -> null
+              is Bulk.Update.Insert -> it.record
+              is Bulk.Update.Modify -> it.record
+          }
+          DbOperation.Insert(record!!)
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/03_implementations.mdx:153]:
```suggestion
          when (it) {
                    Bulk.Prime.Completed -> null
                    is Bulk.Prime.Record -> null
                    is Bulk.Update.Delete -> DbOperation.Delete(it.record)
                    is Bulk.Update.Insert -> DbOperation.Insert(it.record)
                    is Bulk.Update.Modify -> DbOperation.Modify(it.record)
                }
```

File: docs/03_server/09_data-pipelines/03_implementations.mdx [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      }
+      .sink(dbSink())
+}
+```
+
+## Other
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/03_implementations.mdx:159]: Maybe we call this "Handlers"?

File: docs/05_components/doc-management/doc-management-server/doc-mgmt-pipelines.mdx [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+sidebar_label: 'Data Pipelines'
+sidebar_position: 4
+id: doc-management-pipelines
+keywords: [doc-management, pipelines]
+tags:
+  - doc-management
+  - file-storage-pipelines
+---
+
+Here are some implementations provided with the Document Manager that can be used within any data pipeline you have set up. See [Data Pipelines](/server/data-pipelines/introduction/) for more information.
```
Comment (khouari1) [inline docs/05_components/doc-management/doc-management-server/doc-mgmt-pipelines.mdx:12]: We need to call out that you have to add a dependency on file-server-pipelines to access the gpal extension functions / classes

File: docs/05_components/doc-management/doc-management-server/doc-mgmt-pipelines.mdx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        .map {
+            val record = when (it) {
+                Bulk.Prime.Completed -> null
+                is Bulk.Prime.Record -> it.record.toString()
+                is Bulk.Update.Delete -> null
+                is Bulk.Update.Insert -> it.record.toString()
+                is Bulk.Update.Modify -> it.record.toString()
+            }
+        }
+        .sink(record)
```
Comment (khouari1) [inline docs/05_components/doc-management/doc-management-server/doc-mgmt-pipelines.mdx:73]: This is not a valid pipeline example and does not use the file storage sink

File: docs/03_server/09_data-pipelines/04_custom-components.mdx [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+---
+
+### Persistence Manager
+
+There is a persistence manager interface available in the genesis-pipeline-api module which will allow you to decide how you want to manage tracking what you’ve last persisted.
+The interface can be injected directly into your class where the default implementation used is the DbPersistenceManager as mentioned below.
+
+### Database Persistence Manager
+
+The DbPersistenceManager is an implementation of PersistenceManager that stores persistence records in the Async database as a record of the table mentioned below.
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/04_custom-components.mdx:19]:
```suggestion
The DbPersistenceManager is an implementation of PersistenceManager that stores persistence records in the database as a record of the table mentioned below.
```

File: docs/03_server/09_data-pipelines/04_custom-components.mdx [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+This class supports transactions and will use a transaction handle if provided, otherwise will use the injected instance of the entity database.
+The latest index returned will always be the greatest value for the given source. The option is also given to the user whether or not to ‘append’ the record i.e; if true, create a new persistence record. If false, overwrite the current persistence record with the new index value.
+
+As such, there has been a new table created called DATA_PIPELINE_PERSISTENCE which contains the following fields:
+
+- source: String = the source name for which we are storing the persistence information (needs to be unique per pipeline
```
Comment (khouari1) [inline docs/03_server/09_data-pipelines/04_custom-components.mdx:25]:
```suggestion
- source: String = the source name for which we are storing the persistence information (needs to be unique per pipeline)
```

File: docs/05_components/doc-management/doc-management-server/doc-mgmt-pipelines.mdx (hunk 11 above)
Comment (AliAlbaity) [reply docs/05_components/doc-management/doc-management-server/doc-mgmt-pipelines.mdx]: this is now resolved

**Comment by ZuhaaJamani** [comment]
this is now in https://github.com/genesiscommunitysuccess/docs/pull/1783

**Pull request closed without merging**